pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go

```diff
@@ -448,13 +547,33 @@ func (plugin *kubenetNetworkPlugin) teardown(namespace string, name string, id k
 			errList = append(errList, err)
 		}
 	}
+
+	iplist, exists := plugin.podIPs[id]
+	if !exists || iplist.Len() == 0 {
+		klog.V(5).Infof("container %s (%s/%s) does not have recorded. ignoring teardown call", id, name, namespace)
+		return nil
+	}
+
+	for _, ip := range iplist.UnsortedList() {
+		klog.V(5).Infof("Removing pod IP %s from shaper for (%s/%s)", ip, name, namespace)
+		// shaper uses a cidr, but we are using a single IP.
+		isV6 := netutils.IsIPv6String(ip)
+		mask := "32"
+		if isV6 {
+			mask = "128"
+		}
+
+		if err := plugin.shaper().Reset(fmt.Sprintf("%s/%s", ip, mask)); err != nil {
```

Comment:
You have to lock around shaper() or make that function grab the lock  when it initializes plugin.shaper. It's not initialized at plugin init time because we only want to take the hit when somebody acutally does need traffic shaping on a pod.